### PR TITLE
Add AI script action 10104 Chronoshift to Enemy Base

### PR DESCRIPTION
**Pull request by Starkku**
This is basically bastard child of script actions `53 Gather at Enemy Base` and `56/57 Chronoshift to Enemy Building/Target Type` 

Functions for Chronoshifting team to a specific target as well as acquiring the enemy base gather cell are separate so they can potentially be reused in future for something else if needed.

----------------------------------

##### `10104` Chronoshift to Enemy Base

- Chronoshifts the members of the TeamType using first available `Type=Chronosphere` superweapon to a location within `[General]` -> `AISafeDistance` (plus the additional distance defined in parameter, can be negative) cells from enemy house's base. The superweapon must be charged up to atleast `[General]` -> `AIMinorSuperReadyPercent` percentage of its recharge time to be available for use by this action.

In `aimd.ini`:
```ini
[SOMESCRIPTTYPE]  ; ScriptType
x=10104,n         ; integer, additional distance in cells
```

### CODE REVIEW

File: src/Ext/Script/Body.cpp [hunk 1]
```diff
@@ -1177,6 +1182,112 @@ void ScriptExt::Stop_ForceJump_Countdown(TeamClass* pTeam)
...
+	}
+
+	auto const pTargetCell = HouseExt::GetEnemyBaseGatherCell(pEnemy, pLeader->Owner, pLeader->GetCoords(), pLeader->GetTechnoType()->SpeedType, extraDistance);
+
+	if (!pTargetCell)
+	{
+		ScriptExt::Log("AI Scripts - ChronoshiftToEnemyBase: [%s] [%s] (line: %d = %d,%d) Jump to next line: %d = %d,%d -> (Reason: No target cell found)\n", pTeam->Type->ID, pScript->Type->ID, pScript->CurrentMission, pScript->Type->ScriptActions[pScript->CurrentMission].Action, pScript->Type->ScriptActions[pScript->CurrentMission].Argument, pScript->CurrentMission + 1, pScript->Type->ScriptActions[pScript->CurrentMission + 1].Action, pScript->Type->ScriptActions[pScript->CurrentMission + 1].Argument);
+		pTeam->StepCompleted = true;
+		return;
+	}
```
Comment (MortonPL) [inline src/Ext/Script/Body.cpp:1217]: These logging commands are large and repetitve, maybe you could create texts once and use formatting to insert just the reason for failure?

**Review by MortonPL — CHANGES_REQUESTED** [review]

File: YRpp
[the remark this answers is not in the document]
Comment (Starkku) [reply YRpp:1]: Yeah probably. Most if not all applications in the game itself use it as negative, there's two constants for it (one which is the result of this formula, another is 1 divided by it) and both of them are negative too.

File: src/Ext/House/Body.cpp [hunk 2]
```diff
@@ -300,6 +300,38 @@ int HouseExt::TotalHarvesterCount(HouseClass* pThis)
...
+
+	if (currentCoords == CoordStruct::Empty)
+		currentCoords = defaultCurrentCoords;
+
+	int deltaX = currentCoords.X - targetBaseCoords.X;
+	int deltaY = targetBaseCoords.Y - currentCoords.Y;
+	int distance = (RulesClass::Instance->AISafeDistance + extraDistance) * Unsorted::LeptonsPerCell;
+
+	double atan = Math::atan2(deltaY, deltaX);
+	double radians = (((atan - Math::HalfPi) * (1.0 / Math::BinaryAngleMagic)) - 0X3FFF) * Math::BinaryAngleMagic;
```
[the remark this answers is not in the document]
Comment (Starkku) [reply src/Ext/House/Body.cpp:324]: Fair enough on the constant.

It is pretty much just direct port of the reverse-engineered code for action 53. Might be possible to optimize it but I didn't do much simply cause I suck at math.

File: src/Ext/Script/Body.cpp (hunk 1 above)
Comment (Starkku) [reply src/Ext/Script/Body.cpp:1217]: I'll look into it.

**Review by MortonPL — APPROVED** [review]
Looks good, just needs conflict resolution.

**Pull request merged**